app/script/conversation/ConversationTimedHandler.js

```diff
@@ -0,0 +1,221 @@
+/*
+ * Wire
+ * Copyright (C) 2018 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.conversation = z.conversation || {};
+
+z.conversation.ConversationTimedHandler = class ConversationTimedHandler {
+  static get CONFIG() {
+    return {
+      INTERVAL_TIME: 250,
+    };
+  }
+
+  constructor(conversationService, conversationRepository) {
+    this.updateTimedMessages = this.updateTimedMessages.bind(this);
+    this.addTimedMessage = this.addTimedMessage.bind(this);
+    this.removeTimedMessage = this.removeTimedMessage.bind(this);
+    this.checkMessageTimer = this.checkMessageTimer.bind(this);
+
+    this.conversationService = conversationService;
+    this.conversationRepository = conversationRepository;
+    this.logger = new z.util.Logger('z.conversation.ConversationTimedHandler', z.config.LOGGER.OPTIONS);
+
+    this.timedMessages = [];
+    window.setInterval(this.updateTimedMessages, ConversationTimedHandler.CONFIG.INTERVAL_TIME);
+  }
+
+  updateTimedMessages() {
+    const now = Date.now();
+    for (let index = this.timedMessages.length - 1; index >= 0; index--) {
+      const message = this.timedMessages[index];
+      if (_.isString(message.ephemeral_expires())) {
+        const remainingTime = message.ephemeral_expires() - now;
+        if (remainingTime > 0) {
+          message.ephemeral_remaining(remainingTime);
+        } else {
+          this.removeTimedMessage(message);
+        }
+      }
+    }
+  }
+
+  addTimedMessage(message) {
+    if (message.ephemeral_status === z.message.EphemeralStatusType.NONE) {
+      return;
+    }
+    const isAlreadyAdded = this.timedMessages.some(
+      timedMessage => timedMessage.id === message.id && timedMessage.conversation_id === message.conversation_id
+    );
+    if (!isAlreadyAdded) {
+      this.timedMessages.push(message);
+      this.updateTimedMessages();
+    }
+  }
+
+  removeTimedMessage(message) {
+    const messageIndex = this.timedMessages.indexOf(message);
+    if (messageIndex > -1) {
```

Comment:
Let's Filter based on conversation and message IDs.

Reply:
Let's not in this case. I want a specific message object to be removed. There _should_ never be two with colliding conversation and message IDs, but if there by any freaky accident are, i still just want this specific message object out. I underscored the method though, to make it clear, that it should never be called from outside.